src/vdaf/mastic.rs

```diff
@@ -379,32 +382,477 @@ where
+                        verifier_len: Some(verifier_len),
+                    },
+                )
+            } else {
+                (
+                    MasticPrepareShare {
+                        vidpf_proof: xof.into_seed(),
+                        szk_query_share_opt: None,
+                    },
+                    MasticPrepareState {
+                        output_shares: MasticOutputShare::<T::Field>::from(output_shares),
+                        szk_query_state: None,
+                        verifier_len: None,
+                    },
+                )
+            };
+        Ok((prep_state, prep_share))
+    }
+
+    fn prepare_shares_to_prepare_message<
+        M: IntoIterator<Item = MasticPrepareShare<T::Field, SEED_SIZE>>,
+    >(
+        &self,
+        _agg_param: &MasticAggregationParam,
+        inputs: M,
+    ) -> Result<MasticPrepareMessage<T::Field, SEED_SIZE>, VdafError> {
+        let mut inputs_iter = inputs.into_iter();
+        let leader_share = inputs_iter.next().ok_or(VdafError::Uncategorized(
+            "No leader share received".to_string(),
+        ))?;
+        let helper_share = inputs_iter.next().ok_or(VdafError::Uncategorized(
+            "No helper share received".to_string(),
+        ))?;
+        if inputs_iter.next().is_some() {
+            return Err(VdafError::Uncategorized(
+                "more than 2 prepare shares".to_string(),
+            ));
+        };
+
+        match (leader_share, helper_share) {
+            (
+                MasticPrepareShare {
+                    vidpf_proof: leader_vidpf_proof,
+                    szk_query_share_opt: Some(leader_query_share),
+                },
+                MasticPrepareShare {
+                    vidpf_proof: helper_vidpf_proof,
+                    szk_query_share_opt: Some(helper_query_share),
+                },
+            ) => {
+                if leader_vidpf_proof == helper_vidpf_proof {
+                    Ok(Some(SzkQueryShare::merge_verifiers(
+                        leader_query_share,
+                        helper_query_share,
+                    )))
+                } else {
+                    Err(VdafError::Uncategorized(
+                        "Vidpf proof verification failed".to_string(),
+                    ))
+                }
+            }
+            (
+                MasticPrepareShare {
+                    vidpf_proof: leader_vidpf_proof,
+                    szk_query_share_opt: None,
+                },
+                MasticPrepareShare {
+                    vidpf_proof: helper_vidpf_proof,
+                    szk_query_share_opt: None,
+                },
+            ) => {
+                if leader_vidpf_proof == helper_vidpf_proof {
+                    Ok(None)
+                } else {
+                    Err(VdafError::Uncategorized(
+                        "Vidpf proof verification failed".to_string(),
+                    ))
+                }
```

Comment:
Control flow: we check `leader_vidpf_proof != helper_vidpf_proof` twice in this function; it should be possible to refactor so that we only call it once. This is better because there is one and only one place where this step can fail.